### PR TITLE
refactor(DiscordWebsocket): Improvements

### PR DESCRIPTION
**Pull request by SiroTM**
### ws/DiscordWebsocket.ts
+ Added a reason in this.disconnect() for ws.close()
+ Disable the 'close' event in this.disconnect() to work properly
+ Disconnect with a close code and a reason in this.restart()

### ws/constants.ts
+ Gateway CloseEvent codes
+ Removed INVALID_SESSION from GATEWAY_CLOSE_EVENT_CODES

### CODE REVIEW

File: ws/DiscordWebsocket.ts [hunk 1]
```diff
@@ -211,7 +211,7 @@ export default class DiscordWebsocket {
       case GATEWAY_OPCODES.DISPATCH: this._onEvent(p); break;
       case GATEWAY_OPCODES.HEARTBEAT: this._heartbeat(); break;
       case GATEWAY_OPCODES.RECONNECT: this.restart(); break;
-      case GATEWAY_OPCODES.INVALID_SESSION: this._identify(); break;
+      case GATEWAY_OPCODES.INVALID_SESSION: setTimeout(() => this._identify, IDENTIFY_TIMEOUT); break;
```
Comment (Khaaz) [inline ws/DiscordWebsocket.ts:214]: That's fair to wait the identify timeout here eventually?
```suggestion
      case GATEWAY_OPCODES.INVALID_SESSION: setTimeout(this._identify, IDENTIFY_TIMEOUT); break;
```

File: ws/DiscordWebsocket.ts [hunk 2]
```diff
@@ -161,11 +161,11 @@ export default class DiscordWebsocket {
...
       case GATEWAY_CLOSE_EVENT_CODES.UNKNOW_OPCODE:
       case GATEWAY_CLOSE_EVENT_CODES.DECODE_ERROR:
-      case GATEWAY_CLOSE_EVENT_CODES.ALREADY_AUTHENTICATED:
+      case GATEWAY_CLOSE_EVENT_CODES.ALREADY_AUTHENTICATED: this.sessionID = null; break;
       case GATEWAY_CLOSE_EVENT_CODES.RATE_LIMITED: this.restart(); break;
-      case GATEWAY_CLOSE_EVENT_CODES.NOT_AUTHENTICATED:
+      case GATEWAY_CLOSE_EVENT_CODES.NOT_AUTHENTICATED: this.sessionID = null; break;
       case GATEWAY_CLOSE_EVENT_CODES.INVALID_SESSION:
-      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE:
+      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE: this._seq = null; break;
```
Comment (Khaaz) [inline ws/DiscordWebsocket.ts:168]: slightly confused on these 3. Why would we need these special case?

**Review by Khaaz — COMMENTED** [review]
Thanks for the contributions.
Could you try being moer specific in your PR description, listing the changes and maybe including information about why you are doing theses change

File: ws/DiscordWebsocket.ts [hunk 3]
```diff
@@ -161,11 +161,11 @@ export default class DiscordWebsocket {
       case GATEWAY_CLOSE_EVENT_CODES.UNKNOWN_ERROR:
       case GATEWAY_CLOSE_EVENT_CODES.UNKNOW_OPCODE:
       case GATEWAY_CLOSE_EVENT_CODES.DECODE_ERROR:
-      case GATEWAY_CLOSE_EVENT_CODES.ALREADY_AUTHENTICATED:
+      case GATEWAY_CLOSE_EVENT_CODES.ALREADY_AUTHENTICATED: this.sessionID = null; break;
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:164]: Why reset the session?

File: ws/DiscordWebsocket.ts (hunk 1 above)
Comment (bsian03) [reply ws/DiscordWebsocket.ts:214]: Can we not immediately identify?

**Review by bsian03 — COMMENTED** [review]
Will have a proper look through later

File: ws/DiscordWebsocket.ts (hunk 1 above)
Comment (Khaaz) [reply ws/DiscordWebsocket.ts:214]: I guess if we get invalid session straight after an identify?

Comment (SiroTM) [reply ws/DiscordWebsocket.ts:214]: I did not tried to immediately identify but I followed https://discord.com/developers/docs/topics/gateway#resuming-example-gateway-resume

File: ws/DiscordWebsocket.ts (hunk 2 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:168]: INVALID_RESUME_SEQUENCE is when the seq is incorrect, so we have to reset it, no?

File: ws/DiscordWebsocket.ts (hunk 3 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:164]: I don't remember I'll just remove that

File: ws/DiscordWebsocket.ts (hunk 2 above)
Comment (bsian03) [reply ws/DiscordWebsocket.ts:168]: They're all handled by restart with true param

Comment (SiroTM) [reply ws/DiscordWebsocket.ts]: Alright I used that instead

File: ws/DiscordWebsocket.ts [hunk 4]
```diff
@@ -42,21 +42,21 @@ export default class DiscordWebsocket {
...
+    this.ws?.close(code, reason);
     return this;
   }
 
   initialize() {
     if (!this._url) throw new Error('Websocket URL not provided');
 
     this._selfDisconnect = false;
-    if (this.ws) throw new Error('Attempted duplicate connection - run uninitialize');
+    if (this.ws) this.reset();
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:59]: This was intentional, please revert

File: ws/DiscordWebsocket.ts [hunk 5]
```diff
@@ -81,7 +81,8 @@ export default class DiscordWebsocket {
 
   restart(newSession = false) {
     if (newSession) this.sessionID = null;
-    this.disconnect().connect();
+    this.ws?.off('close', this._onClose);
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:84]: Would be better if this was placed in disconnect

File: ws/DiscordWebsocket.ts [hunk 6]
```diff
@@ -165,7 +166,7 @@ export default class DiscordWebsocket {
       case GATEWAY_CLOSE_EVENT_CODES.RATE_LIMITED: this.restart(); break;
       case GATEWAY_CLOSE_EVENT_CODES.NOT_AUTHENTICATED:
       case GATEWAY_CLOSE_EVENT_CODES.INVALID_SESSION:
-      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE:
+      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE: this.restart(true); break;
       case GATEWAY_CLOSE_EVENT_CODES.SESSION_TIMEOUT: setTimeout(() => this.restart(true), IDENTIFY_TIMEOUT); break;
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:170]: This one doesn't need a timeout, that was my bad

File: ws/constants.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // https://discord.com/developers/docs/topics/gateway#commands-and-events-gateway-events
-export const EVENTS = {
+export const EVENTS: { [x: string]: string } = {
```
Comment (bsian03) [inline ws/constants.ts:2]: Leave this for now, I'm planning to make this an enum

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: ws/DiscordWebsocket.ts (hunk 5 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts]: Makes sense

File: ws/DiscordWebsocket.ts [hunk 8]
```diff
@@ -164,9 +165,8 @@ export default class DiscordWebsocket {
       case GATEWAY_CLOSE_EVENT_CODES.ALREADY_AUTHENTICATED:
       case GATEWAY_CLOSE_EVENT_CODES.RATE_LIMITED: this.restart(); break;
       case GATEWAY_CLOSE_EVENT_CODES.NOT_AUTHENTICATED:
-      case GATEWAY_CLOSE_EVENT_CODES.INVALID_SESSION:
-      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE:
-      case GATEWAY_CLOSE_EVENT_CODES.SESSION_TIMEOUT: setTimeout(() => this.restart(true), IDENTIFY_TIMEOUT); break;
+      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE: this.restart(true); break;
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:168]: Unnecessary
```suggestion
      case GATEWAY_CLOSE_EVENT_CODES.INVALID_RESUME_SEQUENCE:
```

File: ws/DiscordWebsocket.ts [hunk 9]
```diff
@@ -42,13 +42,14 @@ export default class DiscordWebsocket {
...
-    if (code === 1000 || code === 1001) {
+  disconnect(code?: GATEWAY_CLOSE_EVENT_CODES, reason?: string) {
+    if (code === GATEWAY_CLOSE_EVENT_CODES.NORMAL || code === GATEWAY_CLOSE_EVENT_CODES.GOING_AWAY) {
       this.sessionID = null;
       this._seq = null;
     }
     this._selfDisconnect = true;
-    this.ws?.close(code);
+    this.ws?.off('close', this._onClose);
+    this.ws?.close(code, reason);
```
[the remark this answers is not in the document]
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:52]: If we reverse them _onClose() would execute and it would think that there was an error
We *want* to disconnect, so we don't need to handle errors, do we?

[the remark this answers is not in the document]
Comment (bsian03) [reply ws/DiscordWebsocket.ts:52]: That's a good point
@Khaaz just double check this please in case either of us are going crazy

**Review by bsian03 — APPROVED** [review]

File: ws/DiscordWebsocket.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Khaaz) [reply ws/DiscordWebsocket.ts:52]: yes that's true, we don't want to loop there. However we might want to call reset on disconnect? Maybe not reset but something that clear all listeners for instance. If we clear only onClose we will have dup onMessage for instance.

File: ws/DiscordWebsocket.ts [hunk 10]
```diff
@@ -42,13 +42,14 @@ export default class DiscordWebsocket {
     return this;
   }
 
-  disconnect(code?: number) {
-    if (code === 1000 || code === 1001) {
+  disconnect(code?: GATEWAY_CLOSE_EVENT_CODES, reason?: string) {
+    if (code === GATEWAY_CLOSE_EVENT_CODES.NORMAL || code === GATEWAY_CLOSE_EVENT_CODES.GOING_AWAY) {
```
[the remark this answers is not in the document]
Comment (bsian03) [reply ws/DiscordWebsocket.ts:46]: This is documented on discord docs, could just link to the relevant location

File: ws/constants.ts [hunk 11]
```diff
@@ -103,23 +103,27 @@ export enum GATEWAY_OPCODES {
   SYNC_CALL,
 }
 
-// https:// discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-close-event-codes
+// https://discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-close-event-codes
+// 1000 and 1001: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
+// 4200 is just a non-1000 close code
 export enum GATEWAY_CLOSE_EVENT_CODES {
+  NORMAL = 1000,
+  GOING_AWAY,
```
[the remark this answers is not in the document]
Comment (bsian03) [reply ws/constants.ts:111]: Harder to maintain. I'm happy with it auto incrementing like this.

[the remark this answers is not in the document]
Comment (Khaaz) [reply ws/constants.ts:111]: fair

File: ws/DiscordWebsocket.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:56]: I tried that and onMessage() still executes one time, it doesn't duplicate

However I noticed that after restarting `{ t: null, s: null, op: 11, d: null }` is being received twice, which means that we are sending the heartbeat twice.  [4128d64](https://github.com/libraryofcode/dis.ts/pull/22/commits/45d919e2eb34b26345fb04e5f3969cc140d9e2bc) fixes that

[the remark this answers is not in the document]
Comment (bsian03) [reply ws/DiscordWebsocket.ts:56]: That seems a bit hacky - In this case I'd rather still reverse them here and add `_onClose` cases for 1000/1001 (also 4200 whilst you're at it)

[the remark this answers is not in the document]
Comment (SiroTM) [reply ws/DiscordWebsocket.ts]: that's better

**Comment by Khaaz** [comment]
is this ready for merge? seems good imo

**Review by bsian03 — CHANGES_REQUESTED** [review]
Have a couple nitpicks, disregard my approval

File: ws/DiscordWebsocket.ts [hunk 12]
```diff
@@ -66,13 +62,14 @@ export default class DiscordWebsocket {
   }
 
   reset() {
-    if (this._heartBeatInterval) clearInterval(this._heartBeatInterval);
+    if (this._heartBeatInterval) {
+      clearInterval(this._heartBeatInterval); this._heartBeatInterval = null;
```
Comment (Khaaz) [inline ws/DiscordWebsocket.ts:66]: why is this one line

**Review by Khaaz — COMMENTED** [review]
actually, I'd need to go through this entirely to be sure. But could we make it slightly clearer regarding what are the different cases possible.
It seems like there are no really any logical reasons why some close events are methods and some part are directly in the switch case. It just feels like we are handling a lot of particular case for all situation which is probably ok but looks a bit messy and hard to follow. Could we think about how we could make it seems like it follows the same execution flow everytime?
Or something about better method names etc so it seems more logical what we are doing in every situation.

File: ws/DiscordWebsocket.ts (hunk 12 above)
Comment (bsian03) [inline ws/DiscordWebsocket.ts:66]:
```suggestion
      clearInterval(this._heartBeatInterval);
      this._heartBeatInterval = null;
```

File: ws/DiscordWebsocket.ts [hunk 13]
```diff
@@ -158,15 +155,23 @@ export default class DiscordWebsocket {
   // TODO Construct proper errors
   private _onClose(code: GATEWAY_CLOSE_EVENT_CODES) {
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:156]: Disable event listener here

File: ws/DiscordWebsocket.ts [hunk 14]
```diff
@@ -184,6 +189,12 @@ export default class DiscordWebsocket {
         this.reset().connect();
         break;
       }
+      case GATEWAY_CLOSE_EVENT_CODES.RECONNECT: {
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:192]: Can we handle reconnect here? Remove the `.disconnect().connect()` when you do this

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: ws/DiscordWebsocket.ts (hunk 14 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:193]: I'm not sure to understand you

File: ws/DiscordWebsocket.ts (hunk 13 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:157]: Doesn't it stop listening when the connexion is closed?

Comment (bsian03) [reply ws/DiscordWebsocket.ts:157]: We get duplicate event listeners if we don't stop listening

File: ws/DiscordWebsocket.ts (hunk 14 above)
Comment (bsian03) [reply ws/DiscordWebsocket.ts:193]: Do the restart in the 4200 close code, this means we don't need to call both disconnect and reconnect in `.restart()` and `.disconnect()` with a 4200 close code will be handled accordingly

File: ws/DiscordWebsocket.ts (hunk 13 above)
Comment (SiroTM) [reply ws/DiscordWebsocket.ts:157]: I disabled the `close` listener

**Comment by Khaaz** [comment]
part for potential refactoring with clearer method name and execution flow (which we should keep in a separate PR for later). Is it good to merge? @bsian03

File: ws/DiscordWebsocket.ts [hunk 15]
```diff
@@ -42,13 +42,9 @@ export default class DiscordWebsocket {
     return this;
   }
 
-  disconnect(code?: number) {
-    if (code === 1000 || code === 1001) {
-      this.sessionID = null;
-      this._seq = null;
-    }
+  disconnect() {
```
Comment (bsian03) [inline ws/DiscordWebsocket.ts:45]: Wait, why get rid of the param? We still want the option of specifying a close code, and all disconnections should happen from here instead of through the ws directly

**Review by bsian03 — CHANGES_REQUESTED** [review]

**Review by bsian03 — APPROVED** [review]

**Review by Khaaz — APPROVED** [review]
